lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/SharedState.hs

```diff
@@ -0,0 +1,357 @@
+    -> SharedStatePending k
+    -> SharedState n k
+mkSharedState accIx state = updateSharedState id $ SharedState
+    { derivationPrefix
+    , fields = PendingFields state
+    }
+  where
+    derivationPrefix = DerivationPrefix (purposeCIP1854, coinTypeAda, accIx)
+
+type SupportsSharedState (n :: NetworkDiscriminant) k =
+    ( MkKeyFingerprint k Address
+    , SoftDerivation k
+    , Typeable n
+    , MkKeyFingerprint k (Proxy n, k 'AddressK XPub)
+    )
+
+-- | Create a new SharedState from root private key and password.
+mkSharedStateFromRootXPrv
+    :: (SupportsSharedState n k, WalletKey k)
+    => (k 'RootK XPrv, Passphrase "encryption")
+    -> Index 'Hardened 'AccountK
+    -> AddressPoolGap
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> SharedState n k
+mkSharedStateFromRootXPrv (rootXPrv, pwd) accIx gap pTemplate dTemplateM =
+    mkSharedState accIx $
+    SharedStatePending accXPub pTemplate dTemplateM gap
+  where
+    accXPub = publicKey $ deriveAccountPrivateKey pwd rootXPrv accIx
+
+-- | Turn a "pending" into an "active" state or identity if already "active"
+updateSharedState
+    :: forall n k. SupportsSharedState n k
+    => (SharedStatePending k -> SharedStatePending k)
+    -> SharedState n k
+    -> SharedState n k
+updateSharedState f st = case fields st of
+    ReadyFields _ -> st
+    PendingFields pending -> case sharedStateFromPending @n (f pending) of
+        Just ready -> st { fields = ReadyFields ready }
+        Nothing -> st { fields = PendingFields (f pending) }
+
+sharedStateFromPending
+    :: forall n k. SupportsSharedState n k
+    => SharedStatePending k
+    -> Maybe (AddressPool 'MultisigScript k)
+sharedStateFromPending (SharedStatePending accXPub pT dT g)
+    | templatesComplete pT dT = Just $
+        mkAddressPool @n (ParentContextMultisigScript accXPub pT dT) g []
+    | otherwise = Nothing
+
+templatesComplete
+    :: ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> Bool
+templatesComplete pTemplate dTemplate =
+    isValid pTemplate && maybe True isValid dTemplate
+  where
+    isValid = isRight . validateScriptTemplate RequiredValidation
+
+-- | The cosigner with his account public key is done per template.
+-- For every template the script is checked if the cosigner is present.
+-- If yes, then the key is inserted. If the key is already present it is going to be
+-- updated. If there is no cosigner present is the script then the cosigner -
+-- account public key map is not changed. The updating works with pending shared state,
+-- and can unpend the shared state if all public keys are present. When already unpended
+-- shared state is used it does not change it.
+addCosignerAccXPub
+    :: (SupportsSharedState n k, WalletKey k)
+    => k 'AccountK XPub
+    -> Cosigner
+    -> CredentialType
+    -> SharedState n k
+    -> SharedState n k
+addCosignerAccXPub accXPub cosigner cred =
+    updateSharedState $ addCosignerAccXPubPending accXPub cosigner cred
```

Comment:
cf comment on Cardano.Wallet.hs module. I think this function would better capture in its return type that it may _fail_ if pre-conditions aren't met.

Reply:
done